File: 2023/10_PipeMaze/pipemaze.py

# ======================================================================
# Pipe Maze
#   Advent of Code 2023 Day 10 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         p i p e m a z e . p y
# ======================================================================
"Pipemaze for the Advent of Code 2023 Day 10 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import re

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
NORTH = (-1, 0)
SOUTH = (1, 0)
EAST = (0, 1)
WEST = (0, -1)
NONE = (0, 0)
START = 'S'
GROUND = '.'
PIPE = '|'

TURNS = {
    NORTH: {
        '|': NORTH,
        '-': NONE,
        'L': NONE,
        'J': NONE,
        '7': WEST,
        'F': EAST,
        '.': NONE,
        'S': NONE,
    },
    SOUTH: {
        '|': SOUTH,
        '-': NONE,
        'L': EAST,
        'J': WEST,
        '7': NONE,
        'F': NONE,
        '.': NONE,
        'S': NONE,
    },
    EAST: {
        '|': NONE,
        '-': EAST,
        'L': NONE,
        'J': NORTH,
        '7': SOUTH,
        'F': NONE,
        '.': NONE,
        'S': NONE,
    },
    WEST: {
        '|': NONE,
        '-': WEST,
        'L': NORTH,
        'J': NONE,
        '7': NONE,
        'F': SOUTH,
        '.': NONE,
        'S': NONE,
    },
}

RE_TO_BLANK = re.compile(r"F-*7|L-*J") # U bends
RE_TO_PIPE = re.compile(r"F-*J|L-*7")  # Kind of pipe -- just shifty

# ======================================================================
#                                                               Pipemaze
# ======================================================================


class Pipemaze(object):   # pylint: disable=R0902, R0903, R0205
    "Object for Pipe Maze"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        self.start = None

        # 2. Process text (if any)
        if text is not None and len(text) > 0:
            self._process_text(text)

    def _process_text(self, text):
        "Assign values from text"

        # 0. Precondition axioms
        assert text is not None and len(text) > 0

        # 1. Find the start
        for rindx, row in enumerate(text):
            sindx = row.find(START)
            if sindx != -1:
                self.start = (rindx, sindx)
                break

    def step(self, location, direction):
        "Return the new location and the new direction"

        # 1. Determine the new location
        new_loc = (location[0] + direction[0], location[1] + direction[1])

        # 2. Determine the pipe at that location
        new_pipe = self.text[new_loc[0]][new_loc[1]]

        # 3. Determine the new direction
        new_direction = TURNS[direction][new_pipe]

        # 4. Return the new direction and location
        return new_loc, new_direction

    def furthest(self):
        "Return the furthest distance from the starting position"

        # 1. Start at the start going in different directions
        animals = [(self.start, SOUTH), (self.start, EAST)]

        # 2. Advance both one step
        animals = [self.step(*animals[0]), self.step(*animals[1])]
        steps = 1

        # 3. Loop until they reach the same location
        while animals[0][0] != animals[1][0]:

            # 4. Advance them both
            animals = [self.step(*animals[0]), self.step(*animals[1])]

            # 5. And increment the step count
            steps += 1

        # 6. Return the furthest distance
        return steps

    def loop_locations(self):
        "Return the locations that are part of the loop"

        # 1. Start at the start and record the location
        animal = (self.start, SOUTH)
        result = set([animal[0]])

        # 2. Advance both one step and record the location
        animal = self.step(*animal)
        result.add(animal[0])

        # 3. Loop until it gets back to the start
        while animal[0] != self.start:

            # 4. Advance one step
            animal = self.step(*animal)

            # 5. And record the pipe segment location
            result.add(animal[0])

        # 6. Return the locations of the pipes in the loop
        return result

    def set_to_ground(self, loop_locs):
        "Turn the now loop squares to ground"

        # 1. Loop for the rows
        for rindx, row in enumerate(self.text):
            row = list(row)

            # 2. Loop for all the columns
            for cindx, _ in enumerate(row):

                # 3. If this square isn't in the loop change it to ground(.)
                if (rindx, cindx) not in loop_locs:
                    row[cindx] = GROUND

            # 4. Update the row
            self.text[rindx] = "".join(row)

    def find_inside(self):
        "Return the count of the inside squares - thanks to ProfONeil's perl regex"

        # 1. Find the squares of the loop pipe
        keep = self.loop_locations()

        # 2. Set the other squares to ground
        self.set_to_ground(keep)

        # 3. Start with nothing
        result = 0

        # 4. Loop for all the rows
        for row in self.text:

            # 5. Get count of the inside squares in this row
            knt = Pipemaze.find_inside_row(row)

            # 6. Update the total count
            result += knt

        # 7. Return the count of inside squares
        return result

    @staticmethod
    def find_inside_row(row):
        "Return the number of inside squares for this row"

        # 1. Use regexs to simplify the row (U bends and offset pipes)
        row = RE_TO_BLANK.sub("", row)
        row = RE_TO_PIPE.sub(PIPE, row)

        # 2. Start with none
        result = 0
        parity = 0

        # 3. Loop for all the remaining squares in the row
        for square in row:

            # 4. If it is a pipe, increment parity
            if square == PIPE:
                parity += 1

            # 5. Else if it is ground, inside if odd parity
            elif square == GROUND:
                if parity % 2 == 1:
                    result += 1

        # 6. Return result
        return result

# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------


if __name__ == '__main__':
    pass

# ======================================================================
# end                      p i p e m a z e . p y                     end
# ======================================================================

File: 2023/10_PipeMaze/pipemaze.go
// ======================================================================
// Pipe Maze
//   Advent of Code 2023 Day 10 -- Eric Wastl -- https://adventofcode.com
//
// go implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                         p i p e m a z e . g o
// ======================================================================
// A solver for the Advent of Code 2023 Day 10 puzzle

package main

// ----------------------------------------------------------------------
//                                                                imports
// ----------------------------------------------------------------------
import (
	"fmt"
	"regexp"
	"slices"
	"strings"
)

// ----------------------------------------------------------------------
//
//	constants
//
// ----------------------------------------------------------------------
const (
	startChar = 'S'
	ground    = '.'
	pipe      = '|'
)

var (
	regToBlank = regexp.MustCompile("F-*7|L-*J|S-*7") //  U bends
	regToPipe  = regexp.MustCompile("F-*J|L-*7|S-*J") //  Kind of pipe -- just shifty
)

// ======================================================================
//                                                               Pipemaze
// ======================================================================

// Pipemaze ... Object for Pipe Maze
type Pipemaze struct {
	Part2 bool
	Text  []string
	Start *Location
}

// NewPipemaze ... Create a Pipemaze object
func NewPipemaze(part2 bool, text []string) (*Pipemaze, error) {

	// 1. Set the initial values
	p := new(Pipemaze)
	p.Part2 = part2
	p.Text = text

	// 2. Process text (if any)
	if len(p.Text) > 0 {
		err := p.processText(p.Text)
		if err != nil {
			return nil, err
		}
	}

	// 3. Return the new Pipemaze object
	return p, nil
}

// processText ... Assign values from text
func (p *Pipemaze) processText(text []string) error {

	// 1. Loop for each line of the text
	for row, line := range text {

		// 2. Loop for each column of the line
		for col, char := range line {

			// 3. Is this the start?
			if char == startChar {
				p.Start = NewLocation(row, col)
				return nil
			}
		}
	}

	// 4. Start not found
	return fmt.Errorf("unable to find starting position")
}

// At ... Return the pipe at the indicated row and column
func (p *Pipemaze) At(row int, col int) rune {

	// 1. Check the coordinates and return pipe if good
	if len(p.Text) > 0 && row >= 0 && col >= 0 && row < len(p.Text) && col < len(p.Text[0]) {
		return rune(p.Text[row][col])
	}

	// 2. Return earth as a default
	fmt.Printf("illegal coordinate (%d, %d)", row, col)
	return ground
}

// Furthest ... Fing the number of steps to the furthest point
func (p *Pipemaze) Furthest() int {

	// 1. Create two animals
	animalOne := NewAnimal(p.Start.Copy(), "East")
	animalTwo := NewAnimal(p.Start.Copy(), "South")

	// 2. Loop until the get to the same place
	for {

		// 3. Let each move
		animalOne.Move(p)
		animalTwo.Move(p)

		// 4. If there are at the same place, we are done
		if animalOne.Loc.Equal(animalTwo.Loc) {
			break
		}
	}

	// 5. Return the number of steps
	return animalOne.Steps
}

// Loop ... Get the locations of the loop in the maze
func (p *Pipemaze) Loop() []*Location {

	// 1. Create an animal to run the maze
	animal := NewAnimal(p.Start.Copy(), "East")

	// 2. Get the locations
	locs := animal.Loop(p)

	// 3. Return the locations
	return locs
}

// OnlyLoop ... Clean up the maze to have only the loop parts
func (p *Pipemaze) OnlyLoop() {

	// 1. Get the locations of the loop in the maze
	locs := p.Loop()

	// 2. Sort the locations so we can use a binary search
	slices.SortFunc(locs, func(a, b *Location) int { return a.SortFunc(b) })

	// 3. Get text that is all ground
	newtext := make([]string, 0, len(p.Text))
	for rowindx := 0; rowindx < len(p.Text); rowindx = rowindx + 1 {
		row := make([]rune, 0, len(p.Text))
		for colindx := 0; colindx < len(p.Text[0]); colindx = colindx + 1 {
			here := NewLocation(rowindx, colindx)
			_, found := slices.BinarySearchFunc(locs, here,
				func(a, b *Location) int { return a.SortFunc(b) })
			athere := ground
			if found {
				athere = p.At(rowindx, colindx)
			}
			row = append(row, athere)
		}
		newtext = append(newtext, string(row))
	}

	// 4. Set the new text
	p.Text = newtext
}

// CountGround ... Count the number of ground spaces
func (p *Pipemaze) CountGround() int {

	// 1. Start with nothing
	result := 0

	// 2. Loop for all the rows
	for indx := 0; indx < len(p.Text); indx = indx + 1 {

		// 3. Get the number of ground locations in this ros
		rowknt := strings.Count(p.Text[indx], string(ground))

		// 4. Add the row count into the total
		result = result + rowknt
	}

	// 5. Return the number of ground spaces
	return result
}

// CountInsideGround ... Count the number of interier ground spaces
func (p *Pipemaze) CountInsideGround() int {

	// 1. Start with nothing
	result := 0

	// 2. Loop for all the rows
	for indx := 0; indx < len(p.Text); indx = indx + 1 {

		// 3. Get the number of interier spaces in this row
		knt := p.CountInsideGroundRow(indx)

		// 4. Accumulte the row counds
		result = result + knt
	}

	// 5. Return the number of ground spaces
	return result
}

// CountInsideGroundRow ... Count the number of interier ground spaces on one row
func (p *Pipemaze) CountInsideGroundRow(rowindx int) int {

	// 1. Start with nothing
	result := 0
	parity := false

	// 2. Get the row text
	row := p.Text[rowindx]

	// 3. Use regexps to simplify the row (U bends and offset pipes)
	row = regToBlank.ReplaceAllLiteralString(row, "")
	row = regToPipe.ReplaceAllLiteralString(row, string(pipe))

	// 5. Loop for all the remaining squares in the row
	for _, square := range row {

		// 6. If at pipe, increment parity
		if square == pipe {
			parity = !parity
		} else {
			// 7. Else if it is inside, increment result
			if square == ground && parity {
				result = result + 1
			}
		}
	}

	// 8. Return result
	return result
}

// PartOne ... Returns the solution for part one
func (p *Pipemaze) PartOne(verbose bool, limit int) string {

	// 1. Return the solution for part one
	return fmt.Sprint(p.Furthest())
}

// PartTwo ... Returns the solution for part two
func (p *Pipemaze) PartTwo(verbose bool, limit int) string {

	// 1. Keep only the pipe loop
	p.OnlyLoop()

	// 2. Count the inside ground nests squares
	result := p.CountInsideGround()

	// 3. Return the solution for part two
	return fmt.Sprint(result)
}

// ======================================================================
// end                      p i p e m a z e . g o                     end
// ======================================================================
